pytriton/proxy/communication.py

```diff
@@ -155,15 +155,19 @@ async def handle_messages(self):
         def _all_responses_processed():
             return not any([self._handle_responses_tasks, self._responses_queues])
 
+        recv_future = None
         try:
             flag_check_interval_s = 1.0
-            # have to receive mssages untill all requestss to be processed, despite shutdown event is set
+            # have to receive messages until all requests are processed, even while shutting down
             while not self._shutdown_event.is_set() or not _all_responses_processed():
+                if recv_future is None:
+                    recv_future = self._socket.recv_multipart()
                 requests_id = b"<unknown>"
                 try:
                     requests_id, flags, responses_payload = await asyncio.wait_for(
-                        self._socket.recv_multipart(), flag_check_interval_s
+                        asyncio.shield(recv_future), flag_check_interval_s
```

Comment:
Suggestion: Switch to zmq.Poller

Using asyncio.shield will disable the periodic check for the shutdown event, which is needed for the graceful closure of the server.

What do you think about using zmq.Poller instead of asyncio.wait_for? It will poll for data readiness on the socket without cancelling the Future, and we will still be able to check the state of the event.

Reply:
Using `asyncio.shield` does *not* disable the periodic check, it will just not cancel the task that receives on the socket when the timeout happens and this avoids dropping messages. Using shield + wait_for is a common pattern, documented [here](https://docs.python.org/3/library/asyncio-task.html#asyncio.wait_for). In #88 I gave [another option](https://github.com/triton-inference-server/pytriton/issues/88#issuecomment-2460252976) using `asyncio.wait` as well.

`zmq.Poller` could probably be used to achieve the same thing but I am not sure it is necessary with asyncio, moreover I am not entirely sure `recv_multipart` is guaranteed not to block when the poller returns (I think the message can be partial).